appinventor/blocklyeditor/src/block_svg.js

```diff
@@ -292,6 +327,44 @@ Blockly.BlockSvg.prototype.renderHere = function(opt_bubble) {
   Blockly.Instrument.stats.renderHereTime += timeDiff;
 };
 
+/**
+ * Makes sure that when the block is collapsed, it is rendered correctly for
+ * that state.
+ * @private
+ */
+Blockly.BlockSvg.prototype.updateCollapsed_ = function() {
+  var collapsed = this.isCollapsed();
+  var collapsedInputName = Blockly.BlockSvg.COLLAPSED_INPUT_NAME;
+  var collapsedFieldName = Blockly.BlockSvg.COLLAPSED_FIELD_NAME;
+
+  for (var i = 0, input; (input = this.inputList[i]); i++) {
+    if (input.name == collapsedInputName) {
+      continue;
+    }
+    input.setVisible(!collapsed);
+  }
+
+  if (!collapsed) {
+    this.removeInput(collapsedInputName);
+    return;
+  }
+
+  var icons = this.getIcons();
+  for (var i = 0, icon; (icon = icons[i]); i++) {
+    icon.setVisible(false);
+  }
+
+  var text = this.toString(Blockly.COLLAPSE_CHARS);
+  var field = this.getField(collapsedFieldName);
+  if (field) {
+    field.setValue(text);
+    return;
+  }
+  var input = this.getInput(collapsedInputName) ||
+      this.appendDummyInput(collapsedInputName);
```

Comment:
Another issue that previous fix uncovered:

![blocks (40)](https://user-images.githubusercontent.com/1331253/77167539-9b330800-6a8c-11ea-85c4-774d4860a019.png)

This is triggered by collapsing a block and reloading the project. During the initial render pass (we only call renderDown on top-level blocks), we get a double render. appendDummyInput adds a new input and immediately tries to render the block we are currently rendering (your this.isRendering check), but because we didn't call render() at the top that value is false when it should be true. Indirectly, that calls this method again, so we end up this input being prepopulated. The following line adds the second field.

Maybe it would make sense to move the isRendering check into renderHere rather than render.

Reply:
@ewpatton So what I ended up doing is reorganizing a bit. I changed `renderDown` and `renderHere` to private functions. This way render is the only access point (invariant). `render` handles all of the "should we render?" checks. The other functions handle all of the logic.

I'm definitely up for changing it, this just felt good to me. So please tell me what you think!

---

I also ended up having to remove renameCollapsed b/c it was causing problems with loading collapsed procedures. Everything seems to be working. I tested all relevant blocks, renaming the relevant parts, and loading from XML. But I will continue testing